pkg/common/jwtutil/keyset.go

```diff
@@ -0,0 +1,141 @@
+package jwtutil
+
+import (
+	"context"
+	"encoding/json"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/zeebo/errs"
+	"gopkg.in/square/go-jose.v2"
+)
+
+type KeySetProvider interface {
+	GetKeySet(context.Context) (*jose.JSONWebKeySet, error)
+}
+
+type KeySetProviderFunc func(context.Context) (*jose.JSONWebKeySet, error)
+
+func (fn KeySetProviderFunc) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, error) {
+	return fn(ctx)
+}
+
+type KeySetConfigURL string
+
+func (c KeySetConfigURL) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, error) {
+	uri, err := FetchKeySetURI(ctx, string(c))
+	if err != nil {
+		return nil, err
+	}
+	return FetchKeySet(ctx, uri)
+}
+
+type CachingKeySetProvider struct {
+	provider        KeySetProvider
+	refreshInterval time.Duration
+
+	mu      sync.Mutex
+	updated time.Time
+	jwks    *jose.JSONWebKeySet
+
+	hooks struct {
+		now func() time.Time
+	}
+}
+
+func NewCachingKeySetProvider(provider KeySetProvider, refreshInterval time.Duration) *CachingKeySetProvider {
+	c := &CachingKeySetProvider{
+		provider:        provider,
+		refreshInterval: refreshInterval,
+	}
+	c.hooks.now = time.Now
+	return c
+}
+
+func (c *CachingKeySetProvider) GetKeySet(ctx context.Context) (*jose.JSONWebKeySet, error) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	now := c.hooks.now()
+
+	if !c.updated.IsZero() && now.Sub(c.updated) < c.refreshInterval {
+		return c.jwks, nil
+	}
+
+	// refresh key set. if there is a failure, log and return the old set if
+	// available.
+	jwks, err := c.provider.GetKeySet(ctx)
+	if err == nil {
+		c.jwks = jwks
+		c.updated = now
+	} else {
+		logrus.Warnf("unable to refresh key set: %v", err)
```

Comment:
If we're gonna log here, we should let CachingKeySetProvider take a logrus.FieldLogger - otherwise, we will lose log config including formatting, log file etc

Reply:
That's kinda tough. This is currently only used by the plugin, which doesn't have a way of getting at the logger being passed through the server internals. If we configured the global logger then it wouldn't be a problem (which I'm a fan of anyway... how often do you have different logging strategies in subcomponents), but even then out-of-proc plugins wouldn't benefit.

Reply:
ok... we can call this out of scope. i know there are difficulties in the out-of-proc plugins... when we moved to builtin i hoped we could solve this but I never got around to it